### ISSUE TITLE
Enable image2image when using "older" models with Mochi v3.2 and newer

### ISSUE THREAD

**Issue by jrittvo** [godly-devotion/MochiDiffusion#255]
An update by Apple to their CoreML package (version 0.4.0) prevents models converted before the update from working with apps built after the update, when using image2image.

The link below is to a repo at Hugging Face that contains model file components that should make image2image fully functional with "older" Stable Diffusion v1.5 derived models when using them with a "newer" app like Mochi Diffusion v3.2 or v4.0.  Models that were converted after April 2023 probably don't need this fix.

Most models available through the CoreML repo at Hugging Face are "older" SD-1.5 derived models that need this fix.

The repo has a single 318 MB zip file to download that will unzip into 5 individual files for placement into existing model folders.  The repo has fairly detailed instructions on use.

https://huggingface.co/jrrjrr/VAEEncoder_Set_For_Image2Image_With_Mochi_Diffusion_v3.2

**Comment by foldfree** [comment]
Just tried it, the fix is working fine for me using `liberty_split-einsum` 
thanks!

**Comment by jrittvo** [comment]
In case the idea of putting a "generic" component into a specific model make anyone wonder, as I understand it, the sole job of the VAEEncoder is to "read" a starting image (and encode it to a series of "noised" steps) so that the image data can be fed to the Unet.  The Unet is where all the real magic happens.  It is the set of "weights" -- the part that -is- specific to a model.  Before there was image2image and ControlNet, models did not need, or use, a VAEEncoder.

**Comment by vzsg** [comment]
That's pretty much correct, the VAEEncoder converts the starting image of img2img/CN into latent space, and then noise gets added to the image, and this is noise sandwich used to feed Unet. In txt2img pipelines, there was only noise.

But VAEEncoder is not really generic, rather that there aren't as many different VAE models[^1] as there are Unet models. You don't usually see this in the Python world, as the checkpoint file either contains one, or a default one is used for the given SD version, making it possible to treat the complete model as just one file[^2].

In Apple's version, they can't stash everything into a single file, since CoreML does not like that, but still their "model" is a complete folder[^3] with all the different models – and that folder is what you select in Mochi.

---

I can totally envision an "advanced" mode where instead of picking a folder (then having Mochi prep it for ControlNet), you could directly select the Unet, VAE, CN models to use to cook the image. This needs support from apple/ml-stable-diffusion as well, but a fork could easily add it.


[^1]: One could argue that among the models, there is one or at most a few different architectures of VAE and Unet, only the weights are different. It's an interesting revelation to compare the same files of different models. But the actual model.mil files are so tiny in comparison to the weights, that breaking mlmodelc-s up is not useful IMHO.
[^2]: Until you actively look for the advanced setting in A1111 to manually select a different VAE.
[^3]: They got weird with the ControlNet update, also cramming those models into the same folder, for little benefit but a lot of wasted space and complications...

**Comment by jrittvo** [comment]
Thanks for the explanation @vzsg.  The secrets inside the black box are slowly getting revealed.

Based on what you say, I should probably add a second set of replacement VAEEncoders derived from an alternate starting VAE.  The set I have posted now was from an SD-1.5 model converted using vae-ft-mse-840000.ckpt  Several anime models suggest using kl-f8-anime.ckpt instead. (Zabriskiji has a great write-up and visual comparison at https://github.com/Zabriskije/CoreML-VAEs).  I'm playing catch up right now after a week's vacation but I'll put this on the To Do list.

To your note 3 above, I wrote somewhere else that it feels like a lot of things in Apple's ml-stable-diffusion get done by one person working something out for themselves. In this case, they were working with one model while developing CN and they threw the CN module models into that model folder.  Quick and easy, meant just feeding the script a single model path.  And when it was all done and worked out, they didn't think about what would make more sense in the real world, with multiple models.

And now that I have you, when are you going to do a PR to add the scheduler you converted in to Mochi?  We established that it was a drop-in replacement for RealESRGAN.mlmodel so you might just need to change the code to use a variable for the mlmodel name in the section that calls RealESRGAN, and then doctor up the Scheduler drop down selection code to match.  It looks like the person who was working on adding Euler A over at apple/ml-stable-diffusion is cleaning up their PR and a merge will happen soon, so the time is ripe.
